cmd/harmony/main.go

```diff
@@ -251,18 +257,100 @@ func createGlobalConfig() *nodeconfig.ConfigType {
 	nodeConfig.SelfPeer = p2p.Peer{IP: *ip, Port: *port, ConsensusPubKey: nodeConfig.ConsensusPubKey}
 
 	nodeConfig.Host, err = p2pimpl.NewHost(&nodeConfig.SelfPeer, nodeConfig.P2pPriKey)
-	if *logConn && nodeConfig.GetNetworkType() != nodeconfig.Mainnet {
-		nodeConfig.Host.GetP2PHost().Network().Notify(utils.NewConnLogger(utils.GetLogInstance()))
-	}
 	if err != nil {
 		panic("unable to new host in harmony")
 	}
+	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Minute)
+	defer cancel()
+	if err := initP2P(ctx, nodeConfig.Host.(*hostv2.HostV2)); err != nil {
+		utils.Logger().Fatal().Err(err).Msg("cannot initialize P2P")
+	}
+
+	if *logConn && nodeConfig.GetNetworkType() != nodeconfig.Mainnet {
+		nodeConfig.Host.GetP2PHost().Network().Notify(utils.NewConnLogger(*utils.Logger()))
+	}
 
 	nodeConfig.DBDir = *dbDir
 
 	return nodeConfig
 }
 
+func initP2P(ctx context.Context, h *hostv2.HostV2) error {
+	// TODO ek – revisit when we increase number of shards.
+	//  We need to start/stop advertising our shard membership across shard
+	//  change, and we need to start/stop discovering shard members as their
+	//  number scales.
+	var bootnodes []peer.AddrInfo
+	for _, ma := range utils.BootNodes {
+		if ai, err := peer.AddrInfoFromP2pAddr(ma); err == nil {
+			bootnodes = append(bootnodes, *ai)
+		} else {
+			utils.Logger().Warn().Err(err).Str("multiaddr", ma.String()).
+				Msg("cannot convert bootnode multiaddr into addrinfo")
+		}
+	}
+	bootLogger := utils.Logger().With().Interface("bootnodes", bootnodes).Logger()
+	bootLogger.Debug().Msg("bootstrapping P2P network")
+	if err := h.Bootstrap(ctx, bootnodes); err != nil {
+		bootLogger.Warn().Err(err).
+			Msg("cannot bootstrap P2P network - expect stalls")
+	}
+	bootLogger.Debug().Msg("bootstrapped P2P network")
+
+	epochSchedule := core.ShardingSchedule.InstanceForEpoch(ethCommon.Big0)
```

Comment:
Assume we restart the network with epoch number not equal 0, we should use that epoch to get epoch schedule. I have modified the setupGenesisAccount to setupInitialAccount function.